gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,190 @@
+// limitations under the License.
+
+package transform2
+
+import (
+	"context"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/api"
+	"github.com/google/gapid/gapis/api/commandGenerator"
+	"github.com/google/gapid/gapis/config"
+)
+
+type transformChain struct {
+	transforms       []Transform
+	out              Writer
+	generator        commandGenerator.CommandGenerator
+	hasBegun         bool
+	hasEnded         bool
+	currentCommandID api.CmdID
+}
+
+// CreateTransformChain creates a transform chain that will run
+// the required transforms on every command coming from commandGenerator
+func CreateTransformChain(generator commandGenerator.CommandGenerator, transforms []Transform, out Writer) *transformChain {
+	return &transformChain{
+		generator:        generator,
+		transforms:       transforms,
+		out:              out,
+		hasBegun:         false,
+		hasEnded:         false,
+		currentCommandID: 0,
+	}
+}
+
+func (chain *transformChain) beginChain(ctx context.Context) error {
+	handleInitialState(chain.out.State())
+
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.BeginTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) endChain(ctx context.Context) error {
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.EndTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "End Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) transformCommand(ctx context.Context, inputCmd api.Cmd, id api.CmdID) error {
+	var err error
+	cmds := []api.Cmd{inputCmd}
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.TransformCommand(ctx, id, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Error on Transform on cmd [%v:%v] with transform [%v] : %v", id, inputCmd, transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, id, cmds, chain.out); err != nil {
```

Comment:
Same as Begin/End chain.

This looks like it could also be parameterised with a lambda.